algorithmic-strategies/dynamic-programming.md

```diff
@@ -0,0 +1,302 @@
+
+### !callout-info
+
+## From Genomics to Git Diffs
+
+Variations of the longest common subsequence problem power solutions for problems ranging from gene sequence comparisons to Git diffs!
+
+### !end-callout
+
+One approach we might take to solve this problem is at each position, to consider the current letter in each string and the remaining portion of each string. If the current characters match, they contribute one matched character count to our total length, plus however many matches there are in the remainders of the strings.
+
+But maybe we can get a better alignment (a longer subsequence) if we don't take the current pair of characters as part of the subsequence. What if we advanced the first string, looking for a better place to start the match. Or what if we advanced the second string? And what if this wasn't a match in the first place? We can advance both strings. For any of these cases, the current characters don't add anything to the overall length of the subsequence. The final length would be the maximum of those three options.
+
+And how do we find the maximum subsequence of the remaining strings? Why with recursion!
+
+![The call tree for finding the longest common subsequence for the strings "ace" and "ae". Because there are three branches at each level, it spreads very quickly. There are large repeated sections of the call tree.](../assets/algorithmic-strategies_dynamic-programming_lcs.png)  
+_Fig. If the recursive explosion of Fibonacci seemed bad, get a load of this! Crossed-out nodes indicate calls that result in a 0 length due to at least one of the input strings being empty. The numbers indicate the maximum value being returned back from a particular branch of execution. Repeated regions are outlined in yellow._
+
+The explosion of calls in this diagram puts Fibonacci to shame! But for small examples, like in the example table above, it can be manageable with a traditional recursive implementation.
+
+```python
+def lcs(str1, str2):
+    if not str1 or not str2:
+        return 0
+
+    # split the first character from the rest
+    first1 = str1[0]
+    rest1 = str1[1:]
+    first2 = str2[0]
+    rest2 = str2[1:]
+
+    # is this spot a match?
+    if first1 == first2:
+        current_score = 1
+    else:
+        current_score = 0
+
+    # the result for this position is the max of the current score
+    # and each of the maxes from the three possibilities:
+    # 1. advance both characters (adding to score if a match)
+    # 2. advance only the first character
+    # 3. advance only the second character
+    result = max(
+        current_score + lcs(rest1, rest2),
+        lcs(rest1, str2),
+        lcs(str1, rest2)
+    )
+
+    return result
+```
+
+But increasing the input size can slow things down quickly! Consider calculating the longest common subsequence for `"tagacgttagtc"` and `"qaqaqgqtqgqc"`. The traditional recursive solution will take a noticeable amount of time.
```

Comment:
Here traditional is referring to it being un-memoized. I was trying to come up with a label other than simple (again simple as opposed to memoized), but simple could leave the reader feeling bad if it doesn't appear simple to them. So I'm looking for a descriptor that distinguishes this as "just the way we might attempt a recursive implementation in the absence of knowledge about memoization." So like calling it an un-memoized version seemed a bit strange, since why would we call this un-memoized if we didn't know that memoization was a thing?